Lms/enable new sales contacts

anathomical opened the pull request:
## WHAT
- Re-enable the routing for the creation of Sales Contact Forms
- Update user auth and password reset logic to account for new `User` records with `sales-contact` roles.
## WHY
We want to use this new workflow for collecting Sales Contact data for incoming sales opportunities, but also want to guarantee two things:
1. If a person submits a sales form request to us and later becomes a user, we want to make sure that the form and user are associated with each other when we sync data to Vitally
2. No one should experience any weirdness with authentication or sign up given that we're using the `User` table in an unconventional way
## HOW
- Stop redirecting sales form requests to the old form
- Confirm that Clever auth acts as it always does even if there is a sales-contact user
- Make sure that password login and password reset treat `sales-contact` `User`s as if they don't exist
- Update Google auth logic to handle `sales-contact` `User`s as if they don't exist unless someone's signing up


### Notion Card Links
https://www.notion.so/quill/Address-edge-cases-in-the-sales-contact-implementation-6bbc884385f348c3af58c4496f6bb3f5

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | Yes
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | N/A
Back-to-school 2022: Have you checked the [webinar schedule](https://www.notion.so/quill/Back-to-school-webinar-banners-2022-a75a89cfad9f434899ef6be3eb184733) to avoid for downtime/risky deploys? | N/A

anathomical committed:
Re-enable the new sales contact form and Intercom webhooks

anathomical committed:
WIP attempts to get tests for new Google expectations to work

anathomical committed:
Update Google auth code to handle `sales-contact` users

Basically, if you try to "sign in" and have a sales-user record, treat it as if you don't have a user at all (and show an error).  If you "sign up" and you have a sales-user record, update the `role` on that record to the appropriate type based on sign-up.

anathomical committed:
Update password reset to treat non_authenticating Users as missing

We have a specific message that we use when a user tries to re-set a password for a "non-existing" user, and we want to use that same message if the user has a non-authenticating user type like "sales-contact" since they don't know they have an account and we want them to go through the sign-up process.

anathomical committed:
Treat non-authenticating user roles as non-existant during sign-in

We want a user who signs in with email/password for an email that is assigned to a `sales-concat` user type to get the same experience as someone signing in with an email that doesn't correspond to any user.  This is because we want to direct these users to the sign-up flow if they actually want to use Quill.

anathomical committed:
Fix linting errors

anathomical committed:
Simplify some early return logic to avoid linting errors

dandrabik commented on services/QuillLMS/app/models/user.rb:
```diff
@@ -662,6 +663,15 @@ def segment_user
+  # through our onboarding flow, but are given a User record so that we
+  # can sync their data to Vitally.  We need to treat these users specially
+  # during auth flows because they haven't actually signed up.
+  def non_authenticating?
```
I think using a positive name here, e.g.  `.prospective?` would make some of the code easier to read and avoid double/triple negatives, e.g. `!@user.non_authenticating` or `return unless @user.non_authenticating` 

And I understand what you are trying to do here with the naming and this comment, but you should also consider just naming this `sales_contact?`. We only have one type in the role, it might be better to keep the naming specific until we have a second case.

anathomical replied:
I went back and forth on what to call this, and sort of went with the broader version "just in case", but I think you're right that we might as well keep the scope of this restricted until we actually know we have another example.  There's a good chance we'd want to refactor this if we had additional use cases anyway.

dandrabik commented on services/QuillLMS/app/controllers/sessions_controller.rb:
```diff
@@ -46,7 +46,7 @@ def create
     email_or_username = params[:user][:email].downcase.strip unless params[:user][:email].nil?
     @user =  User.find_by_username_or_email(email_or_username)
-    if @user.nil?
+    if @user.nil? || @user&.non_authenticating?
```
You don't need a second `&.` nil check here.

dandrabik commented on services/QuillLMS/app/controllers/auth/google_controller.rb:
```diff
@@ -85,7 +85,28 @@ def online_access_callback
+    session[:role].present?
+  end
+
+  private def update_role_if_necessary
```
This is a very general name with a specific purpose. Renaming/ordering might make this easier to read, e.g.:
```ruby
def update_sales_contact_role
  return unless @user.sales_contact? && in_sign_up_flow?

dandrabik reviewed (changes requested):
Thanks for pushing this along. I think some re-naming here might make some of the code easier to read, so I'm requesting that, e.g. something like `sales_contact?` instead of `non_authenticating?` 

I'm also a little worried that we are adding another edge-case in code that has a lot of branching logic, but I understand this is kinda a means-to-an-end for now. We'll just have to monitor and be on top of potential bugs.

anathomical committed:
Update code based on review feedback

anathomical committed:
After discussion and testing with Peter S, disable this link again

We'll turn it back on once we figure out with Vitally what we actually want to do around managing the receipt of these forms

dandrabik reviewed (approved):
Thanks for making these changes!
